src/core/qgspointlocator.cpp

```diff
@@ -343,7 +343,9 @@ static QgsPointLocator::MatchList _geometrySegmentsInRect( QgsGeometry *geom, co
 
   QgsWkbTypes::Type wkbType = geom->wkbType();
 
-  bool hasZValue = false;
+  bool hasZValue = QgsWkbTypes::hasZ( wkbType );
+  bool hasMValue = QgsWkbTypes::hasM( wkbType );
+
   switch ( wkbType )
```

Comment:
Looks like this would be better suited to either switch on QgsWkbTypes::flatType( wkbType ), OR (better?) remove the default clause so that warnings are thrown and EVERY wkb type must be explicitly handled

Reply:
I think we can simplify by converting geometries:
- by deleting Z and M
- by segmenting the curves
- by transforming polygons into multilinestring
There are only linestring or multilinestring left. We could use only the [multistring code](https://github.com/qgis/QGIS/blob/master/src/core/qgspointlocator.cpp#L398-L427) and remove the case statement.
But I think it has a cost in terms of indexing time

Reply:
@nyalldawson https://github.com/qgis/QGIS/pull/9379/commits/c6a86ee70ec71b06aa1766e1f365893d276b259d is the application of my previous comment